### PR TITLE
add CLI step framework

### PR DESCRIPTION
**Pull request by kalensk**
This creates a framework to separate out UI concerns from running CLI or hub steps.

[Pipeline](https://cm.ci.gpdb.pivotal.io/teams/main/pipelines/gpupgrade:refactorCLISteps)

### CODE REVIEW

**Comment by bhuvnesh2703** [comment]
for all the cli step i believe now we see 2 lines: i.e IN PROGRESS and COMPLETE. But IN PROGRESS should be overwritten with COMPLETE. I like the overall approach though and makes it consistent. But this should be fixed.
command: gpupgrade initialize -f gpupgrade_config
```
Validating source and target versions...                           [IN PROGRESS]
Validating source and target versions...                           [COMPLETE]
Creating directories...                                            [IN PROGRESS]
Creating directories...                                            [COMPLETE]
Creating cluster configuration file...                             [IN PROGRESS]
Creating cluster configuration file...                             [COMPLETE]
Starting gpupgrade hub process...                                  [IN PROGRESS]
Starting gpupgrade hub process...                                  [COMPLETE]
Saving source cluster configuration...                             [COMPLETE]
Starting gpupgrade agent processes...                              [COMPLETE]
Checking disk space...                                             [IN PROGRESS]
Checking disk space...                                             [COMPLETE]
Generating target cluster configuration...                         [COMPLETE]
Creating target cluster...                                         [COMPLETE]
Stopping target cluster...                                         [COMPLETE]
Backing up target master...                                        [COMPLETE]
Running pg_upgrade checks...                                       [COMPLETE]

Initialize completed successfully.
```

vs
before
```
Creating directories...                                            [COMPLETE]
Starting gpupgrade hub process...                                  [COMPLETE]
Saving source cluster configuration...                             [COMPLETE]
Starting gpupgrade agent processes...                              [COMPLETE]
Checking disk space...                                             [COMPLETE]
Generating target cluster configuration...                         [COMPLETE]
Creating target cluster...                                         [COMPLETE]
Stopping target cluster...                                         [COMPLETE]
Backing up target master...                                        [COMPLETE]
Running pg_upgrade checks...                                       [COMPLETE]
```

**Comment by kalensk** [comment]
> for all the cli step i believe now we see 2 lines: i.e IN PROGRESS and COMPLETE. But IN PROGRESS should be overwritten with COMPLETE. I like the overall approach though and makes it consistent. But this should be fixed.
command: gpupgrade initialize -f gpupgrade_config

Thanks! I pushed a fixup that addresses this.

File: cli/commanders/steps.go [hunk 1]
```diff
@@ -74,7 +74,7 @@ func InitializeCreateCluster(client idl.CliToHubClient, verbose bool) (err error
 	return nil
 }
 
-func Execute(client idl.CliToHubClient, verbose bool) (*ExecuteResponse, error) {
```
Comment (dkrieger-pvtl) [inline cli/commanders/steps.go:77]: The only change in this file is to change the return type of the first returned argument from a pointer to a value.  Why was this change made?  It also seems unrelated to this PR.

File: cli/errors.go [hunk 2]
```diff
@@ -3,7 +3,24 @@
...
 package cli
 
-import "fmt"
+import (
+	"errors"
+	"fmt"
+)
+
+// ErrNextActions is an error type that allows customizing the NextActions text.
+type ErrNextActions struct {
```
Comment (dkrieger-pvtl) [inline cli/errors.go:12]: This type is introduced for a single usage(for version checking failing in initialize), but the existing NextActions already includes an error as part of its type.  I am unclear why it was introduced.  It would make sense as part of a later refactor that prints out a NextActions generically for each step, but it's not clear to me we'd choose to introduce a new type there.

**Review by dkrieger-pvtl — COMMENTED** [review]
I've given it a thorough first pass.  I like the overall approach.  However, the addition of changes that seem unrelated to the main idea(changing pointers to references in the return values, adding a new ErrNextActions type) complicates the review.

File: cli/commanders/steps.go (hunk 1 above)
Comment (kalensk) [reply cli/commanders/steps.go:77]: Yeah, one of the tests was failing during this PR. See the commit message for details.


> user commanders Response instance rather than pointer
>
> Don't use a pointer for commanders.ExecuteResponse,
commanders.FinalizeResponse, and commanders.RevertResponse since they may be
null if connecting to the hub fails. This will result in a null pointer as the
response is used when step.Complete is executed.
>
> This was causing "subcommands return an error if the hub is not started" in
test/initialize.bats to fail with an exit status of 2 instead of 1.

**Comment by bhuvnesh2703** [comment]
Now the issue of printing IN PROGRESS and COMPLETE Is resolved. The rest looks good to me, but we should figure out the UI aspect of this `Validating source and target versions...                           [COMPLETE]`. should it be shown or not?

File: cli/errors.go (hunk 2 above)
Comment (kalensk) [reply cli/errors.go:12]: I wanted to preserve the existing behavior where `ValidateVersions` would not show the revert suggestion in next actions. What about renaming `ErrNextActions` to `SetNextActions` or similar would be clearer?

File: cli/commanders/steps.go [hunk 3]
```diff
@@ -83,60 +83,60 @@ func Execute(client idl.CliToHubClient, verbose bool) (*ExecuteResponse, error)
...
 			TargetMasterDataDir: response[idl.ResponseKey_target_master_data_directory.String()],
 		},
 		nil
 }
 
-func Revert(client idl.CliToHubClient, verbose bool) (*RevertResponse, error) {
+func Revert(client idl.CliToHubClient, verbose bool) (RevertResponse, error) {
 	fmt.Println()
 	fmt.Println("Revert in progress.")
 	fmt.Println()
```
[the remark this answers is not in the document]
Comment (kalensk) [reply cli/commanders/steps.go]: Updated. Thanks!

File: cli/commands/commands.go [hunk 4]
```diff
@@ -622,46 +585,33 @@ func revert() *cobra.Command {
...
-			s.Finish(&err)
-			if err != nil {
-				return err
-			}
+			st.RunCLISubstep(idl.Substep_STOP_HUB_AND_AGENTS, func(streams step.OutStreams) error {
+				return stopHubAndAgents(false)
+			})
 
-			commanders.LogDuration(operation, verbose, timer.Stop())
+			st.RunCLISubstep(idl.Substep_STOP_HUB_AND_AGENTS, func(streams step.OutStreams) error {
```
[the remark this answers is not in the document]
Comment (kalensk) [reply cli/commands/commands.go]: Updated. Thanks!

File: cli/commanders/steps.go [hunk 5]
```diff
@@ -83,60 +83,60 @@ func Execute(client idl.CliToHubClient, verbose bool) (*ExecuteResponse, error)
...
 		TargetPort:          response[idl.ResponseKey_target_port.String()],
 		TargetMasterDataDir: response[idl.ResponseKey_target_master_data_directory.String()],
 	}, nil
 }
 
-func Finalize(client idl.CliToHubClient, verbose bool) (*FinalizeResponse, error) {
+func Finalize(client idl.CliToHubClient, verbose bool) (FinalizeResponse, error) {
 	fmt.Println()
 	fmt.Println("Finalize in progress.")
 	fmt.Println()
```
[the remark this answers is not in the document]
Comment (kalensk) [reply cli/commanders/steps.go]: Updated. Thanks!

File: cli/commanders/steps.go [hunk 6]
```diff
@@ -74,7 +74,7 @@ func InitializeCreateCluster(client idl.CliToHubClient, verbose bool) (err error
 	return nil
 }
 
-func Execute(client idl.CliToHubClient, verbose bool) (*ExecuteResponse, error) {
+func Execute(client idl.CliToHubClient, verbose bool) (ExecuteResponse, error) {
 	fmt.Println()
 	fmt.Println("Execute in progress.")
 	fmt.Println()
```
[the remark this answers is not in the document]
Comment (kalensk) [reply cli/commanders/steps.go]: Updated. Thanks!

**Comment by kalensk** [comment]
I'll update the PR to not show the following substeps on the UI. I "may" create a followup PR or story if it ends up being a bit too much work, but hopefully not.

- Substep_CREATE_CLUSTER_CONFIG
- Substep_VALIDATE_VERSIONS

File: cli/commands/commands.go [hunk 7]
```diff
@@ -393,96 +394,70 @@ func initialize() *cobra.Command {
...
+			})
+
+			st.RunCLISubstep(idl.Substep_CREATING_DIRECTORIES, func(streams step.OutStreams) error {
+				return commanders.CreateStateDir()
+			})
 
-			fmt.Println()
-			fmt.Println("Initialize in progress.")
-			fmt.Println()
+			st.RunCLISubstep(idl.Substep_CREATE_CLUSTER_CONFIG, func(streams step.OutStreams) error {
```
[the remark this answers is not in the document]
Comment (kalensk) [reply cli/commands/commands.go]: This commit has been removed with the addition of `RunInternalSubstep`

**Comment by kalensk** [comment]
Take a final look. I think I addressed all comments. I added `RunInternalSubstep` that does not display anything to the UI.

**Review by bhuvnesh2703 — APPROVED** [review]
Sorry forgot that the comment was pending..

**Comment by kalensk** [comment]
> Also, should Substep_CREATING_DIRECTORIES and Substep_CREATE_CLUSTER_CONFIG be under a common step as they are preparatory type of actions?

You mean have this:
```go
st.RunCLISubstep(idl.Substep_CREATING_DIRECTORIES, func(streams step.OutStreams) error {
        err := commanders.CreateStateDir()
        if err != nil {
            return err
        }

        err = commanders.CreateInitialClusterConfigs(hubPort)
        if err != nil {
            return err
        }

        return nil
    })
```

Instead of:
```go
st.RunCLISubstep(idl.Substep_CREATING_DIRECTORIES, func(streams step.OutStreams) error {
    return commanders.CreateStateDir()
})

st.RunInternalSubstep(func() error {
    return commanders.CreateInitialClusterConfigs(hubPort)
})
```

File: cli/commanders/step.go [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+	"golang.org/x/xerrors"
+
+	"github.com/greenplum-db/gpupgrade/cli"
+	"github.com/greenplum-db/gpupgrade/idl"
+	"github.com/greenplum-db/gpupgrade/step"
+	"github.com/greenplum-db/gpupgrade/utils/errorlist"
+	"github.com/greenplum-db/gpupgrade/utils/stopwatch"
+)
+
+type ClIStep struct {
```
[the remark this answers is not in the document]
Comment (kalensk) [reply cli/commanders/step.go]: Yes! Thank you. Fixed.

File: cli/errors.go (hunk 2 above)
Comment (kalensk) [reply cli/errors.go]: Added `func (s *CLIStep) SetNextActions(suggestRevert bool)` which looks like:
```go
st.RunInternalSubstep(func() error {
err := cli.ValidateVersions(sourceGPHome, targetGPHome)
if err != nil {
    st.SetNextActions(false)
    return err
}

return nil
})
```

File: cli/commands/commands.go [hunk 9]
```diff
@@ -393,96 +394,70 @@ func initialize() *cobra.Command {
...
-					// XXX work around an annoying UI nit: the error message is
-					// smashed against the failing substep without any vertical
-					// space. Ideally this would be handled by the "UI" logic as
-					// opposed to pushed into business logic.
-					fmt.Println()
+			st := commanders.NewStep(idl.Step_INITIALIZE, &step.BufferedStreams{}, verbose)
 
-					err = cli.NewNextActions(err, "initialize", suggestRevert)
+			// ValidateVersions before starting the hub and running the rest of initialize.
+			st.RunCLISubstep(idl.Substep_VALIDATE_VERSIONS, func(streams step.OutStreams) error {
```
[the remark this answers is not in the document]
Comment (kalensk) [reply cli/commands/commands.go]: The substep has been updated to use `RunInternalSubstep`

**Review by dkrieger-pvtl — APPROVED** [review]
LGTM.  This is pretty clean and makes the business logic(flow of initialize, execute, finalize and revert) easier to see. Note that, unlike hub/steps, all CLISteps are always re-run and are assumed to be idempotent.  This is fine for now, though I do wonder if we want to call `RunCLISubstep` as `AlwaysRunCLISubstep` as otherwise the reader(us!) needs to keep this in mind.
